File: runtime/kernel/kernel_bridge.R
#!/usr/bin/env Rscript
# Minimal local R kernel for the Workbench notebook.
#
# A persistent process that holds one environment across cells (shared state,
# like a Jupyter kernel). The host writes the cell's code to a file (argv[1]),
# then sends one line "<id>" on stdin; we evaluate that file in the global
# environment and write ONE JSON response line back to stdout:
#
#     response: {"id","ok","stdout","result","error"}\n
#
# Base R only — no jsonlite/IRkernel — so it runs against any R install,
# offline, with no model key. `result` mirrors Jupyter: the printed value of the
# cell's final expression when it is visible, else null.

args <- commandArgs(trailingOnly = TRUE)
codefile <- args[1]
options(warn = 1) # surface warnings inline (stdout), not deferred to session end

# JSON-escape a scalar string. fixed = TRUE keeps replacements literal so we do
# not fight regex metacharacters; control chars left after \n\r\t are dropped.
json_escape <- function(s) {
  if (length(s) != 1) s <- paste(s, collapse = "\n")
  if (is.na(s)) return("")
  s <- gsub("\\", "\\\\", s, fixed = TRUE)
  s <- gsub("\"", "\\\"", s, fixed = TRUE)
  s <- gsub("\n", "\\n", s, fixed = TRUE)
  s <- gsub("\r", "\\r", s, fixed = TRUE)
  s <- gsub("\t", "\\t", s, fixed = TRUE)
  s <- gsub("[[:cntrl:]]", "", s)
  s
}

emit <- function(id, ok, out, result, error) {
  parts <- c(
    paste0("\"id\":\"", json_escape(id), "\""),
    paste0("\"ok\":", if (ok) "true" else "false"),
    paste0("\"stdout\":\"", json_escape(out), "\""),
    if (is.null(result)) "\"result\":null" else paste0("\"result\":\"", json_escape(result), "\""),
    if (is.null(error)) "\"error\":null" else paste0("\"error\":\"", json_escape(error), "\"")
  )
  cat(paste0("{", paste(parts, collapse = ","), "}"), "\n", sep = "")
  flush(stdout())
}

run_cell <- function(code) {
  exprs <- tryCatch(parse(text = code), error = function(e) e)
  if (inherits(exprs, "error")) {
    return(list(ok = FALSE, stdout = "", result = NULL,
                error = paste0("Error: ", conditionMessage(exprs))))
  }
  captured <- character(0)
  buf <- textConnection("captured", open = "w", local = TRUE)
  sink(buf)
  sink(buf, type = "message")
  result <- NULL
  err <- NULL
  tryCatch({
    n <- length(exprs)
    if (n > 0) for (i in seq_len(n)) {
      wv <- withVisible(eval(exprs[[i]], envir = globalenv()))
      if (wv$visible) {
        printed <- paste(utils::capture.output(print(wv$value)), collapse = "\n")
        if (i == n) result <- printed else cat(printed, "\n", sep = "")
      }
    }
  }, error = function(e) {
    err <<- paste0("Error: ", conditionMessage(e))
  })
  sink(type = "message")
  sink()
  close(buf)
  list(ok = is.null(err), stdout = paste(captured, collapse = "\n"),
       result = result, error = err)
}

con <- file("stdin", open = "r")
repeat {
  line <- readLines(con, n = 1)
  if (length(line) == 0) break # host closed stdin -> exit
  id <- trimws(line)
  if (nchar(id) == 0) next
  code <- tryCatch(paste(readLines(codefile, warn = FALSE), collapse = "\n"),
                   error = function(e) "")
  r <- run_cell(code)
  emit(id, r$ok, r$stdout, r$result, r$error)
}

File: runtime/kernel/kernel_bridge.py
#!/usr/bin/env python3
"""Minimal local Python kernel for the Workbench notebook.

A persistent process that holds one namespace across cells (shared state, like a
Jupyter kernel) and speaks a line-delimited JSON protocol over stdin/stdout:

    request : {"id": "<str>", "code": "<str>"}\\n
    response: {"id","ok","stdout","result","error"}\\n

Standard library only — no ipykernel/ZMQ — so it runs against whatever Python the
user has, offline, with no model key. `result` mirrors Jupyter: the repr of the
final expression when a cell ends in one, else null.
"""
import ast
import io
import json
import sys
import traceback
from contextlib import redirect_stderr, redirect_stdout


def run_cell(ns: dict, code: str):
    """Execute `code` in namespace `ns`. Returns (stdout, result_repr_or_None, error_or_None)."""
    out = io.StringIO()
    try:
        parsed = ast.parse(code, mode="exec")
    except SyntaxError:
        return "", None, traceback.format_exc(limit=1)

    body = parsed.body
    result = None
    # Jupyter behaviour: if the cell ends in an expression, show its value.
    tail_expr = None
    if body and isinstance(body[-1], ast.Expr):
        last = body.pop()
        assert isinstance(last, ast.Expr)
        tail_expr = ast.Expression(last.value)

    try:
        with redirect_stdout(out), redirect_stderr(out):
            if body:
                exec(compile(ast.Module(body, []), "<cell>", "exec"), ns)  # noqa: S102
            if tail_expr is not None:
                value = eval(compile(tail_expr, "<cell>", "eval"), ns)  # noqa: S307
                if value is not None:
                    result = repr(value)
    except Exception:  # surface the traceback to the notebook, like a kernel does
        return out.getvalue(), None, traceback.format_exc()

    return out.getvalue(), result, None


def main() -> None:
    ns: dict = {"__name__": "__main__"}
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            req = json.loads(line)
        except json.JSONDecodeError:
            continue
        stdout, result, error = run_cell(ns, req.get("code", ""))
        resp = {
            "id": req.get("id"),
            "ok": error is None,
            "stdout": stdout,
            "result": result,
            "error": error,
        }
        sys.stdout.write(json.dumps(resp) + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()
